PowerShellTools/DebugEngine/ScriptDebugger.cs

```diff
@@ -208,139 +195,192 @@ private void HandleRemoteSessionForwardedEvent(object sender, PSEventArgs args)
+        {
+            Log.InfoFormat("SetBreakpoint: {0} {1} {2}", breakpoint.File, breakpoint.Line, breakpoint.Column);
 
             try
             {
-                using (var pipeline = (_runspace.CreatePipeline()))
-                {
-                    var command = new Command("Remove-PSBreakpoint");
-                    command.Parameters.Add("Breakpoint", breakpoints);
-                    pipeline.Commands.Add(command);
-
-                    pipeline.Invoke();
-                }
+                _debuggingService.SetBreakpoint(new PowershellBreakpoint(breakpoint.File, breakpoint.Line, breakpoint.Column));
             }
             catch (Exception ex)
             {
-                Log.Error("Failed to clear breakpoints.", ex);
+                Log.Error("Failed to set breakpoint.", ex);
             }
-
         }
+        
+        #region Debugging service event handlers
 
-        private void SetBreakpoint(ScriptBreakpoint breakpoint)
+
+
+
+        /// <summary>
+        /// Debugger stopped handler
+        /// </summary>
+        /// <param name="e"></param>
+        public void DebuggerStop(DebuggerStoppedEventArgs e)
         {
-            Log.InfoFormat("SetBreakpoint: {0} {1} {2}", breakpoint.File, breakpoint.Line, breakpoint.Column);
+            Log.InfoFormat("Debugger stopped");
 
-            try
+            RefreshScopedVariables();
+            RefreshCallStack();
+
+            if (!ProcessLineBreakpoints(e.ScriptFullPath, e.Line, e.Column))
             {
-                using (var pipeline = (_runspace.CreatePipeline()))
+                if (DebuggerPaused != null)
                 {
-                    var command = new Command("Set-PSBreakpoint");
-                    command.Parameters.Add("Script", breakpoint.File);
-                    command.Parameters.Add("Line", breakpoint.Line);
+                    var scriptLocation = new ScriptLocation();
+                    scriptLocation.File = e.ScriptFullPath;
+                    scriptLocation.Line = e.Line;
+                    scriptLocation.Column = 0;
 
-                    pipeline.Commands.Add(command);
+                    DebuggerPaused(this, new EventArgs<ScriptLocation>(scriptLocation));
+                }
+            }
 
-                    pipeline.Invoke();
+            Log.Debug("Waiting for debuggee to resume.");
+
+            //Wait for the user to step, continue or stop
+            _pausedEvent.WaitOne();
+            Log.DebugFormat("Debuggee resume action is {0}", _resumeAction);
+
+            _debuggingService.SetResumeAction(_resumeAction);
+        }
+
+        /// <summary>
+        /// Output string from debugger in VS output/REPL pane window
+        /// </summary>
+        /// <param name="output"></param>
+        public void VsOutputString(string output)
+        {
+            if (ReplWindow != null)
+            {
+                if (output.StartsWith("[ERROR]"))

```

Comment:
Fixed.

From: Huanhuan Sun [mailto:notifications@github.com]
Sent: Wednesday, January 28, 2015 1:35 PM
To: Microsoft/poshtools
Cc: Eric Deng
Subject: Re: [poshtools] Outproc(64bit) local debugging (#62)


In PowerShellTools/DebugEngine/ScriptDebugger.cs<https://github.com/Microsoft/poshtools/pull/62#discussion_r23725807>:

> +            //Wait for the user to step, continue or stop

> +            _pausedEvent.WaitOne();

> +            Log.DebugFormat("Debuggee resume action is {0}", _resumeAction);

> +

> +            _debuggingService.SetResumeAction(_resumeAction);

> +        }

> +

> +        /// <summary>

> +        /// Output string from debugger in VS output/REPL pane window

> +        /// </summary>

> +        /// <param name="output"></param>

> +        public void VsOutputString(string output)

> +        {

> +            if (ReplWindow != null)

> +            {

> +                if (output.StartsWith("[ERROR]"))

This one seems like a common constant. Suggests put it in a constant file.

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/poshtools/pull/62/files#r23725807>.